### PR TITLE
Bluetooth: controller: Adding user ext. anchor point move to ctrl/CPR

### PR DESCRIPTION
**Pull request by erbr-ot**
Adding support for allowing user extension code to support deferring
anchor point moves. Refactored LLCP only.
Note: This is NOT supported by LEGACY LLCP impl.
Porting central handling of anchor point move and preferred periodicity to refactored LLCP
KConfig'd by BT_CTLR_USER_CPR_ANCHOR_POINT_MOVE
Modified ll_conn_update API, to accommodate passing offsets
Modified CPR/CU context data structure
Modified CPR/CU procedure to handle anchor point move wait state
Fixed up unit tests, adding anchor point move cases to CPR/CU test

Signed-off-by: Erik Brockhoff <erbr@oticon.com>

### CODE REVIEW

File: subsys/bluetooth/controller/Kconfig.ll_sw_split [hunk 1]
```diff
@@ -815,6 +815,13 @@ config BT_CTLR_USER_CPR_INTERVAL_MIN
 	  When enabled, controller will accept Connection Parameter Request
 	  intervals down to a proprietary minimum value.
 
+config BT_CTLR_USER_CPR_ANCHOR_POINT_MOVE
+	bool "Proprietary handling of anchor point movement"
```
Comment (mtpr-ot) [inline subsys/bluetooth/controller/Kconfig.ll_sw_split:819]: No longer proprietary...

Comment (erbr-ot) [reply subsys/bluetooth/controller/Kconfig.ll_sw_split:819]: Well, the USER_EXT is still referred to as 'proprietary extensions in controller', so I guess using proprietary term here is as valid as for the other ones?

**Comment by erbr-ot** [comment]
Revisiting, reorganizing, ...

File: subsys/bluetooth/controller/ll_sw/ull_llcp.h [hunk 2]
```diff
@@ -152,6 +152,18 @@ uint8_t ull_cp_remote_dle_pending(struct ll_conn *conn);
...
+ * @brief Check if a remote connection param reg is expecting an
+ *        anchor point move response.
+ */
+bool ull_cp_remote_cpr_awaiting_apm_response(struct ll_conn *conn);
+
+/**
+ * @brief Repsond to anhor point move of remote connection param
+ *        reg.
+ */
+void ull_cp_remote_cpr_apm_respond(struct ll_conn *conn, uint16_t *offsets);
```
Comment (thoh-ot) [inline subsys/bluetooth/controller/ll_sw/ull_llcp.h:165]: I would suggest that the API is changed to

`ull_cp_remote_cpr_apm_reply()`
`ull_cp_remote_cpr_apm_neg_reply()`

such that it resembles other API's that deal with accepting or rejecting procedures.

File: subsys/bluetooth/controller/ll_sw/ull_llcp.h [hunk 3]
```diff
@@ -152,6 +152,18 @@ uint8_t ull_cp_remote_dle_pending(struct ll_conn *conn);
...
 uint8_t ull_cp_remote_cpr_pending(struct ll_conn *conn);
 
+/**
+ * @brief Check if a remote connection param reg is expecting an
+ *        anchor point move response.
+ */
+bool ull_cp_remote_cpr_awaiting_apm_response(struct ll_conn *conn);
+
+/**
+ * @brief Repsond to anhor point move of remote connection param
```
Comment (thoh-ot) [inline subsys/bluetooth/controller/ll_sw/ull_llcp.h:162]: Spelling mistake in 'anhor'

File: subsys/bluetooth/controller/Kconfig.ll_sw_split [hunk 4]
```diff
@@ -815,6 +815,13 @@ config BT_CTLR_USER_CPR_INTERVAL_MIN
...
+	depends on BT_CTLR_USER_EXT && BT_CTLR_CONN_PARAM_REQ
+	help
+	  When enabled, defer accept/reject of Connection Parameter Request with only
+	  anchor point movement to proprietary logic.
+
```
Comment (thoh-ot) [inline subsys/bluetooth/controller/Kconfig.ll_sw_split:824]: As far as i can see the accept/reject of anchor point movement is only for the peripheral, so shouldn't that be mentioned and be dependent on it.

**Review by thoh-ot — CHANGES_REQUESTED** [review]

File: subsys/bluetooth/controller/ll_sw/ull_llcp_conn_upd.c [hunk 5]
```diff
@@ -139,6 +140,11 @@ enum {
 
 	/* CONN_PARAM_REQ negative reply */
 	RP_CU_EVT_CONN_PARAM_REQ_NEG_REPLY,
+
+#if defined(CONFIG_BT_CTLR_USER_CPR_ANCHOR_POINT_MOVE)
```
[the remark this answers is not in the document]
Comment (erbr-ot) [reply subsys/bluetooth/controller/ll_sw/ull_llcp_conn_upd.c:144]: Yeap

File: subsys/bluetooth/controller/ll_sw/ull_llcp.h (hunk 2 above)
Comment (erbr-ot) [reply subsys/bluetooth/controller/ll_sw/ull_llcp.h:165]: Check

**Review by thoh-ot — CHANGES_REQUESTED** [review]

**Comment by erbr-ot** [comment]
Rebased, but cte unit test still fails.

**Review by thoh-ot — DISMISSED** [review]

**Comment by erbr-ot** [comment]
Rebased

**Comment by erbr-ot** [comment]
rebased

**Comment by erbr-ot** [comment]
rebase

**Comment by erbr-ot** [comment]
rebase

**Review by thoh-ot — APPROVED** [review]

**Review by kruithofa — APPROVED** [review]
LGTM

**Pull request merged**